### PR TITLE
C critical section implementation for POSIX platforms

### PR DESCRIPTION
**Pull request by bogdanm**
Pretty much a copy-paste from the C++ code in CriticalSectionLock.h

NOTE:  core-util doesn't compile at this moment for POSIX platforms, but
this particular file does (using target x86-linux-native).

### CODE REVIEW

**Comment by bogdanm** [comment]
@rgrover @bremoran

File: source/critical.c [hunk 1]
```diff
@@ -15,7 +15,9 @@
  * limitations under the License.
  */
 
-#ifndef TARGET_NORDIC
+// This critical section implementation is generic for mbed OS targets,
+// except Nordic ones
+#if defined(TARGET_LIKE_MBED) && !defined(TARGET_NORDIC)
```
Comment (0xc0170) [inline source/critical.c]: This changeset in this file should be a separate commit

Comment (bogdanm) [inline source/critical.c]: Not really. This is linked to the POSIX platform, since the previous condition (`#ifndef TARGET_NORDIC`) assumed that the only possibilities are "regular" mbed OS and Nordic platforms. This change acknowledges the presence of another platform (POSIX).

**Comment by rgrover** [comment]
+1

**Comment by bremoran** [comment]
LGTM
